### PR TITLE
net: Support partial checksum offloading

### PR DESCRIPTION
**Pull request by WorldofJARcraft**
The Xilinx AXI Ethernet subsystem is capable of RX/Tx checksum offloading. While it supports computing IP and UDP/TCP checksums, it does not support computing ICMP checksums and only computes IP checksums for ICMP messages. Thus, this patch adds an default-disabled option for always computing the ICMP checksum in software while relying on hardware to compute the remaining checksums.

### CODE REVIEW

**Review by pdgendt — CHANGES_REQUESTED** [review]
This should be a driver capability, not a global setting. If multiple drivers are used, each should work properly.

File: subsys/net/ip/Kconfig [hunk 1]
```diff
@@ -899,6 +899,14 @@ config NET_DISABLE_ICMP_DESTINATION_UNREACHABLE
 	  Suppress the generation of ICMP destination unreachable errors
 	  when ports that are not in a listening state receive packets.
 
+config NET_FORCE_ICMP_CHECKSUM
```
Comment (tbursztyka) [inline subsys/net/ip/Kconfig:902]: The issue with such approach is as soon as you mix 2 hardware that have different support, i.e. one which has full cksum support, the other only partially . it will unilaterally shut the capabilities of the first one. 

Perhaps it is time to consider a finer ETHERNET_HW_<TX/RX>_CHKSUM_OFFLOAD set of options: one for ipv4/6, icmp4/6?

Comment (WorldofJARcraft) [reply subsys/net/ip/Kconfig:902]: I would be happy to implement either of these options and submit a new PR, but I am not sure what the best solution is.

Comment (rlubos) [reply subsys/net/ip/Kconfig:902]: Perhaps we could leave capabilities as they are, but allow to provide more detailed information about checksum offloading with `get_config()` and a new enum indeed? That should have minimal impact on existing drivers, as in case the more detailed information is not available, we could just retain the current behavior (all protos offloaded).

@tbursztyka WDYT?

Comment (tbursztyka) [reply subsys/net/ip/Kconfig:902]: Sure. The hw enum would be then for a generic support, and the config one for detailed ones. Fine by me

Comment (WorldofJARcraft) [reply subsys/net/ip/Kconfig:902]: Sounds good, I will implement this accordingly and update this PR.

Comment (WorldofJARcraft) [reply subsys/net/ip/Kconfig]: I have added the corresponding new flags. If you are happy with the changes, I will implement the new flag for the existing in-tree ethernet drivers. As it stands now, they would otherwise default to computing all checksums in software.

File: drivers/mdio/Kconfig [hunk 2]
```diff
@@ -1,49 +0,0 @@
-# MDIO configuration options
```
Comment (rlubos) [inline drivers/mdio/Kconfig:1]: Removing this file doesn't seem intentional?

File: include/zephyr/net/ethernet.h [hunk 3]
```diff
@@ -467,6 +469,24 @@ struct ethernet_txtime_param {
...
+enum ethernet_checksum_support {
+	ETHERNET_CHECKSUM_SUPPORT_NONE		  = 0,
+	ETHERNET_CHECKSUM_SUPPORT_IPV4_HEADER = BIT( 0),
+	ETHERNET_CHECKSUM_SUPPORT_IPV4_TCP    = BIT( 1),
+	ETHERNET_CHECKSUM_SUPPORT_IPV4_UDP    = BIT( 2),
+	ETHERNET_CHECKSUM_SUPPORT_IPV4_ICMP   = BIT( 3),
+	/** Space for future protocols and restrictions for IPV4 */
+	ETHERNET_CHECKSUM_SUPPORT_IPV6_HEADER = BIT(16),
+	ETHERNET_CHECKSUM_SUPPORT_IPV6_TCP    = BIT(17),
+	ETHERNET_CHECKSUM_SUPPORT_IPV6_UDP    = BIT(18),
```
Comment (rlubos) [inline include/zephyr/net/ethernet.h:482]: Do we need to differentiate IPv4/6 TCP/UDP? I'd stick to the originally proposed list: IPv4/6, TCP, UDP, ICMPv4/6.

File: include/zephyr/net/net_if.h [hunk 4]
```diff
@@ -2643,6 +2643,21 @@ void net_if_unregister_link_cb(struct net_if_link_cb *link);
 void net_if_call_link_cb(struct net_if *iface, struct net_linkaddr *lladdr,
 			 int status);
 
+/**
+ * @brief Type of checksum for which support in the interface will be queried.
+ */
+enum net_if_checksum_protocol_type{
+	NET_IF_CHECKSUM_PROTOCOL_UNKNOWN,
+	NET_IF_CHECKSUM_PROTOCOL_IPV4_HEADER,
```
Comment (rlubos) [inline include/zephyr/net/net_if.h:2651]: Ditto

File: subsys/net/ip/CMakeLists.txt [hunk 5]
```diff
@@ -64,4 +64,4 @@ zephyr_library_include_directories(. ${ZEPHYR_BASE}/subsys/net/lib)
 zephyr_library_include_directories(
   ${ZEPHYR_BASE}/kernel/include
   ${ARCH_DIR}/${ARCH}/include
-  )
+  )
```
Comment (rlubos) [inline subsys/net/ip/CMakeLists.txt:67]: Doesn't seem intentional.

File: subsys/net/ip/ipv4.c [hunk 6]
```diff
@@ -135,7 +135,11 @@ int net_ipv4_finalize(struct net_pkt *pkt, uint8_t next_header_proto)
 		   next_header_proto == IPPROTO_TCP) {
 		return net_tcp_finalize(pkt, false);
 	} else if (next_header_proto == IPPROTO_ICMP) {
+#ifdef CONFIG_NET_FORCE_ICMP_CHECKSUM
```
Comment (rlubos) [inline subsys/net/ip/ipv4.c:138]: This change should be dropped.

File: include/zephyr/net/net_if.h [hunk 7]
```diff
@@ -2643,6 +2643,21 @@ void net_if_unregister_link_cb(struct net_if_link_cb *link);
 void net_if_call_link_cb(struct net_if *iface, struct net_linkaddr *lladdr,
 			 int status);
 
+/**
+ * @brief Type of checksum for which support in the interface will be queried.
+ */
+enum net_if_checksum_protocol_type{
```
Comment (rlubos) [inline include/zephyr/net/net_if.h:2649]: Actually, looking at the code this enum name looks a bit lengthy ,maybe just call this `net_if_checksum_type` and `NET_IF_CHECKSUM_IPV4` etc...

File: subsys/net/ip/tcp.c [hunk 8]
```diff
@@ -4051,7 +4065,7 @@ int net_tcp_finalize(struct net_pkt *pkt, bool force_chksum)
 
 	tcp_hdr->chksum = 0U;
 
-	if (net_if_need_calc_tx_checksum(net_pkt_iface(pkt)) || force_chksum) {
+	if (net_if_need_calc_tx_checksum(net_pkt_iface(pkt), is_ipv6 ? NET_IF_CHECKSUM_PROTOCOL_IPV6_TCP : NET_IF_CHECKSUM_PROTOCOL_IPV4_TCP) || force_chksum) {
```
Comment (rlubos) [inline subsys/net/ip/tcp.c:4068]: As mentioned above, IMO there's no need to overcomplicate with separate IPv4/IPv6 configs for TCP/UDP.

File: subsys/net/ip/net_if.c [hunk 9]
```diff
@@ -4932,14 +4932,58 @@ void net_if_call_link_cb(struct net_if *iface, struct net_linkaddr *lladdr,
...
+
 	if (net_if_l2(iface) != &NET_L2_GET_NAME(ETHERNET)) {
 		return true;
 	}
 
-	return !(net_eth_get_hw_capabilities(iface) & caps);
+	have_checksum_support = net_eth_get_hw_capabilities(iface) & caps;
+	err = net_eth_get_hw_config(iface,caps == ETHERNET_HW_RX_CHKSUM_OFFLOAD ? ETHERNET_CONFIG_TYPE_RX_CHECKSUM_SUPPORT : ETHERNET_CONFIG_TYPE_TX_CHECKSUM_SUPPORT, &config);
+	if(err){
+		NET_DBG("Could not get ethernet config: %u",-err);
```
Comment (rlubos) [inline subsys/net/ip/net_if.c:4952]: Let's not print anything here, `get_config()` is actually not widely implemented, it will make the log output unreadable for many drivers.

File: subsys/net/ip/net_if.c [hunk 10]
```diff
@@ -4932,14 +4932,58 @@ void net_if_call_link_cb(struct net_if *iface, struct net_linkaddr *lladdr,
...
+		return true;
```
Comment (rlubos) [inline subsys/net/ip/net_if.c:4953]: And actually, I'd revert the logic here. So in case driver does not provide any extra information about HW checksum support other than capability, assume every protocol is supported. That way we maintain backward compatibility, w/o a need to modify all of the drivers.

What I had in mind:
```c
if (!(net_eth_get_hw_capabilities(iface) & caps)) {
	return true; /* No checksum offload*/
}

if (net_eth_get_hw_config(iface, type, &config) != 0) {
	return false; /* No extra info, assume all offloaded. */
}

/* Protocol-specific check. */ 
...
```

File: subsys/net/ip/net_if.c [hunk 11]
```diff
@@ -4932,14 +4932,58 @@ void net_if_call_link_cb(struct net_if *iface, struct net_linkaddr *lladdr,
...
+
+	have_checksum_support_protocol = false;
+
 	if (net_if_l2(iface) != &NET_L2_GET_NAME(ETHERNET)) {
 		return true;
 	}
 
-	return !(net_eth_get_hw_capabilities(iface) & caps);
+	have_checksum_support = net_eth_get_hw_capabilities(iface) & caps;
+	err = net_eth_get_hw_config(iface,caps == ETHERNET_HW_RX_CHKSUM_OFFLOAD ? ETHERNET_CONFIG_TYPE_RX_CHECKSUM_SUPPORT : ETHERNET_CONFIG_TYPE_TX_CHECKSUM_SUPPORT, &config);
```
Comment (rlubos) [inline subsys/net/ip/net_if.c:4950]: `caps == ETHERNET_HW_RX_CHKSUM_OFFLOAD ? ETHERNET_CONFIG_TYPE_RX_CHECKSUM_SUPPORT : ETHERNET_CONFIG_TYPE_TX_CHECKSUM_SUPPORT` - I'd add a helper variable for this, it reads hard when inside a function call.

File: subsys/net/ip/net_if.c [hunk 12]
```diff
@@ -4932,14 +4932,58 @@ void net_if_call_link_cb(struct net_if *iface, struct net_linkaddr *lladdr,
...
+	if(err){
+		NET_DBG("Could not get ethernet config: %u",-err);
+		return true;
+	}
+
+	switch(chksum_type){
+		case NET_IF_CHECKSUM_PROTOCOL_IPV4_HEADER:
```
Comment (rlubos) [inline subsys/net/ip/net_if.c:4957]: Actually, not sure if we need a separate enum for net_if and ethernet, it just brings extra complexity to the solution.
I'd consider defining `enum net_if_checksum_type` as a bitmap at the net_if level, and just use it in ethernet layer. 
That way we could simply `return !(config.chksum_support & chksum_type);`

**Comment by rlubos** [comment]
Please check compliance results, there are missing spaces etc.

File: drivers/mdio/Kconfig (hunk 2 above)
Comment (WorldofJARcraft) [reply drivers/mdio/Kconfig:1]: Yes, I will restore the original file.

File: subsys/net/ip/CMakeLists.txt (hunk 5 above)
Comment (WorldofJARcraft) [reply subsys/net/ip/CMakeLists.txt:67]: Yes, I will restore it.

File: include/zephyr/net/net_if.h (hunk 7 above)
Comment (WorldofJARcraft) [reply include/zephyr/net/net_if.h:2649]: OK, I will change it accordingly.

File: include/zephyr/net/net_if.h (hunk 4 above)
Comment (WorldofJARcraft) [reply include/zephyr/net/net_if.h:2651]: OK

File: include/zephyr/net/ethernet.h (hunk 3 above)
Comment (WorldofJARcraft) [reply include/zephyr/net/ethernet.h:482]: OK, I will merge the UDP and TCP for IPv4/6.

File: subsys/net/ip/ipv4.c (hunk 6 above)
Comment (WorldofJARcraft) [reply subsys/net/ip/ipv4.c:138]: OK

File: subsys/net/ip/net_if.c (hunk 9 above)
Comment (WorldofJARcraft) [reply subsys/net/ip/net_if.c:4952]: OK

File: subsys/net/ip/net_if.c (hunk 10 above)
Comment (WorldofJARcraft) [reply subsys/net/ip/net_if.c:4953]: OK, implemented accordingly.

File: subsys/net/ip/net_if.c (hunk 11 above)
Comment (WorldofJARcraft) [reply subsys/net/ip/net_if.c:4950]: OK, done.

File: subsys/net/ip/net_if.c (hunk 12 above)
Comment (WorldofJARcraft) [reply subsys/net/ip/net_if.c:4957]: See my comment above - it needs to be checked that both the IP and the transport protocol are supported.

Comment (rlubos) [reply subsys/net/ip/net_if.c]: Ok, but we still could have a single enum (bitmap) for net_if and ethernet so we wouldn't need to translate here?

File: subsys/net/ip/tcp.c [hunk 13]
```diff
@@ -4043,6 +4043,20 @@ int net_tcp_finalize(struct net_pkt *pkt, bool force_chksum)
 {
 	NET_PKT_DATA_ACCESS_DEFINE(tcp_access, struct net_tcp_hdr);
 	struct net_tcp_hdr *tcp_hdr;
+	bool is_ipv6;
```
Comment (rlubos) [inline subsys/net/ip/tcp.c:4046]: This could be simplified to a single line:
```c
enum net_if_checksum_type type = net_pkt_family(pkt) == AF_INET6 ?  
	NET_IF_CHECKSUM_IPV6_TCP : NET_IF_CHECKSUM_IPV4_TCP;
```
and the use `type` as an argument in the function. Same in similar other places.

**Review by rlubos — COMMENTED** [review]
Please check the compliance report.
Also the commit title could be updated, the changeset is no longer specific to ICMP.

**Comment by decsny** [comment]
This is convenient, I was planning to look into something like this after LTS because NXP ENET has the same limitation, it only doesn't accelerate ICMPv6 checksum

File: subsys/net/ip/tcp.c (hunk 13 above)
Comment (WorldofJARcraft) [reply subsys/net/ip/tcp.c:4046]: Good idea, done.

**Comment by WorldofJARcraft** [comment]
> Please check the compliance report. Also the commit title could be updated, the changeset is no longer specific to ICMP.

Some of the complaints in the compliance report refer to code that I did not change. I could run the corresponding files through clang-format or manually fix the violations, but this would bloat the PR.

File: subsys/net/ip/net_if.c (hunk 12 above)
Comment (WorldofJARcraft) [reply subsys/net/ip/net_if.c]: I have implemented a version of this idea.
Because ethernet.h includes net_if.h, I needed to add a few extra defines to make this work, however.

File: include/zephyr/net/ethernet.h [hunk 14]
```diff
@@ -913,6 +934,29 @@ enum ethernet_hw_caps net_eth_get_hw_capabilities(struct net_if *iface)
...
+ * @brief Return ethernet device hardware configuration information.
+ *
+ * @param iface Network interface
+ * @param config Ethernet configuration
+ *
+ * @return 0 if ok, <0 if error
+ */
+static inline
+enum ethernet_hw_caps net_eth_get_hw_config(struct net_if *iface, enum ethernet_config_type type,
+struct ethernet_config *config)
```
Comment (rlubos) [inline include/zephyr/net/ethernet.h:947]: nit: Please align with the opening bracket.

File: subsys/net/ip/net_if.c [hunk 15]
```diff
@@ -4932,14 +4932,71 @@ void net_if_call_link_cb(struct net_if *iface, struct net_linkaddr *lladdr,
 	k_mutex_unlock(&lock);
 }
 
-static bool need_calc_checksum(struct net_if *iface, enum ethernet_hw_caps caps)
+static bool need_calc_checksum(struct net_if *iface, enum ethernet_hw_caps caps,
+	enum net_if_checksum_type chksum_type)
```
Comment (rlubos) [inline subsys/net/ip/net_if.c:4936]: nit: Please align with the opening bracket.

File: include/zephyr/net/net_if.h [hunk 16]
```diff
@@ -2643,6 +2643,35 @@ void net_if_unregister_link_cb(struct net_if_link_cb *link);
...
+#define NET_IF_CHECKSUM_IPV6_ICMP_BIT			BIT(11)
+#define NET_IF_CHECKSUM_TCP_BIT				BIT(21)
+#define NET_IF_CHECKSUM_UDP_BIT				BIT(22)
+
+/**
+ * @brief Type of checksum for which support in the interface will be queried.
+ */
+enum net_if_checksum_type {
+	NET_IF_CHECKSUM_IPV4_HEADER = NET_IF_CHECKSUM_IPV4_HEADER_BIT,
+	NET_IF_CHECKSUM_IPV4_TCP = NET_IF_CHECKSUM_IPV4_HEADER_BIT|
```
Comment (rlubos) [inline include/zephyr/net/net_if.h:2662]: I'm not sure I follow, why check for IPV4_HEADER here? I'm not sure what the actual hardware support in the matter, but those doesn't seem related, L4/L3 checksum calculation should IMO be independent.

Comment (WorldofJARcraft) [reply include/zephyr/net/net_if.h:2662]: This is more important for IPv6 - if a network device only supports IPv4 and TCP checksums (in IPv4 packets), then we have to compute IPv6 checksums in software.
The check for IPv4 support could theoretically be omitted (I am not aware of any devices that support IPv6 checksums but do not support IPv4).

Comment (WorldofJARcraft) [reply include/zephyr/net/net_if.h:2662]: I am not sure if this answers your question, but L3 support is required for L4 support. A NIC that does not understand the L3 protocol has no idea where the L4 packet begins and ends, what transport protocol it is and where its checksum field would be. Thus, it cannot insert an L4 checksum.

File: include/zephyr/net/ethernet.h (hunk 14 above)
Comment (WorldofJARcraft) [reply include/zephyr/net/ethernet.h]: Done. Btw., the return type was also incorrect, this is fixed as well.

File: subsys/net/ip/net_if.c (hunk 15 above)
Comment (WorldofJARcraft) [reply subsys/net/ip/net_if.c]: Done

File: include/zephyr/net/net_if.h (hunk 16 above)
Comment (rlubos) [reply include/zephyr/net/net_if.h:2662]: Ok, I see your point, fine by me then.

File: include/zephyr/net/net_if.h [hunk 17]
```diff
@@ -2643,6 +2643,35 @@ void net_if_unregister_link_cb(struct net_if_link_cb *link);
...
+		NET_IF_CHECKSUM_TCP_BIT,
+	NET_IF_CHECKSUM_IPV4_UDP = NET_IF_CHECKSUM_IPV4_HEADER_BIT|
+		NET_IF_CHECKSUM_UDP_BIT,
```
Comment (rlubos) [inline include/zephyr/net/net_if.h:2665]:
```suggestion
	NET_IF_CHECKSUM_IPV4_UDP = NET_IF_CHECKSUM_IPV4_HEADER_BIT |
				   NET_IF_CHECKSUM_UDP_BIT,
```
etc, just reads better.

**Comment by rlubos** [comment]
> This is convenient, I was planning to look into something like this after LTS because NXP ENET has the same limitation, it only doesn't accelerate ICMPv6 checksum

@decsny Good to hear that - feel free to participate in the review, to make sure those changes fit your HW needs as well.

File: include/zephyr/net/net_if.h (hunk 17 above)
Comment (WorldofJARcraft) [reply include/zephyr/net/net_if.h]: Done

**Comment by WorldofJARcraft** [comment]
> > This is convenient, I was planning to look into something like this after LTS because NXP ENET has the same limitation, it only doesn't accelerate ICMPv6 checksum
> 
> @decsny Good to hear that - feel free to participate in the review, to make sure those changes fit your HW needs as well.

I have a draft implementation for the Xilinx AXI Ethernet subsystem, which is where I initially came across this issue, here: https://github.com/zephyrproject-rtos/zephyr/pull/73986

File: include/zephyr/net/net_if.h [hunk 18]
```diff
@@ -2643,16 +2643,47 @@ void net_if_unregister_link_cb(struct net_if_link_cb *link);
 void net_if_call_link_cb(struct net_if *iface, struct net_linkaddr *lladdr,
 			 int status);
 
+/* used to ensure encoding of checksum support in net_if.h and
```
Comment (jukkar) [inline include/zephyr/net/net_if.h:2646]: Doxygen comments missing, if these comments should be internal then add INTERNAL_HIDDEN guards

File: include/zephyr/net/net_if.h [hunk 19]
```diff
@@ -2643,16 +2643,47 @@ void net_if_unregister_link_cb(struct net_if_link_cb *link);
...
+#define NET_IF_CHECKSUM_IPV6_ICMP_BIT			BIT(11)
+#define NET_IF_CHECKSUM_TCP_BIT				BIT(21)
+#define NET_IF_CHECKSUM_UDP_BIT				BIT(22)
+
+/**
+ * @brief Type of checksum for which support in the interface will be queried.
+ */
+enum net_if_checksum_type {
+	NET_IF_CHECKSUM_IPV4_HEADER = NET_IF_CHECKSUM_IPV4_HEADER_BIT,
+	NET_IF_CHECKSUM_IPV4_TCP    = NET_IF_CHECKSUM_IPV4_HEADER_BIT |
```
Comment (tbursztyka) [inline include/zephyr/net/net_if.h:2662]: Is such granularity required?
TCP or UDP are IP agnostic (unlike ICMP). Hardware really have different support like that for l4 protocols, depending on l3?

**Review by tbursztyka — CHANGES_REQUESTED** [review]

File: include/zephyr/net/net_if.h (hunk 19 above)
Comment (WorldofJARcraft) [reply include/zephyr/net/net_if.h:2662]: IPv4 and IPv6 have different header sizes. Also, for both protocols, the header size is variable and depends on the options selected (e.g., IPv6 has a number of optional headers that sit between the mandatory header and the transport header). The transport protocol used is also encoded in the IP header.
Thus, a NIC that does not understand the L3 protocol does not know which transport protocol is used, where the transport packet begins and where it ends and cannot insert a checksum for TCP or UDP.

File: include/zephyr/net/net_if.h (hunk 18 above)
Comment (WorldofJARcraft) [reply include/zephyr/net/net_if.h:2646]: I will add INTERNAL_HIDDEN guards; these definitions are only used to ensure the values of the enums in net_if.h and ethernet.h match.

File: include/zephyr/net/net_if.h [hunk 20]
```diff
@@ -2643,16 +2643,47 @@ void net_if_unregister_link_cb(struct net_if_link_cb *link);
...
  * in which case the IP stack does not need to calculate the checksum.
  *
  * @param iface Network interface
+ * @param chksum_type L3 and/or L4 protocol for which to compute checksum
  *
  * @return True if checksum needs to be calculated, false otherwise.
  */
-bool net_if_need_calc_rx_checksum(struct net_if *iface);
+bool net_if_need_calc_rx_checksum(struct net_if *iface,
+	enum net_if_checksum_type chksum_type);
```
[the remark this answers is not in the document]
Comment (WorldofJARcraft) [reply include/zephyr/net/net_if.h]: Done

File: include/zephyr/net/net_if.h [hunk 21]
```diff
@@ -2661,10 +2692,12 @@ bool net_if_need_calc_rx_checksum(struct net_if *iface);
  * checksum.
  *
  * @param iface Network interface
+ * @param chksum_type L3 and/or L4 protocol for which to compute checksum
  *
  * @return True if checksum needs to be calculated, false otherwise.
  */
-bool net_if_need_calc_tx_checksum(struct net_if *iface);
+bool net_if_need_calc_tx_checksum(struct net_if *iface,
+	enum net_if_checksum_type chksum_type);
```
[the remark this answers is not in the document]
Comment (WorldofJARcraft) [reply include/zephyr/net/net_if.h]: Done

File: include/zephyr/net/net_if.h (hunk 18 above)
Comment (WorldofJARcraft) [reply include/zephyr/net/net_if.h:2648]: I have added doxygen comments for the actual enum values.

**Review by tbursztyka — DISMISSED** [review]

**Review by jukkar — DISMISSED** [review]

File: include/zephyr/net/ethernet.h [hunk 22]
```diff
@@ -465,6 +467,26 @@ struct ethernet_txtime_param {
...
 
+/** Protocols that are supported by checksum offloading */
+enum ethernet_checksum_support {
+	/** Device does not support any L3/L4 checksum offloading */
+	ETHERNET_CHECKSUM_SUPPORT_NONE			= NET_IF_CHECKSUM_NONE_BIT,
+	/** Device supports checksum offloading for the IPv4 header */
+	ETHERNET_CHECKSUM_SUPPORT_IPV4_HEADER		= NET_IF_CHECKSUM_IPV4_HEADER_BIT,
+	/** Device supports checksum offloading for ICMPv4 payload (implies IPv4 header) */
+	ETHERNET_CHECKSUM_SUPPORT_IPV4_ICMP		= NET_IF_CHECKSUM_IPV4_ICMP_BIT,
+	/* Space for future protocols and restrictions for IPV4 */
```
Comment (pdgendt) [inline include/zephyr/net/ethernet.h:478]: This comment doesn't really fit here anymore, should probably be moved to the bits.

File: subsys/net/ip/net_if.c [hunk 23]
```diff
@@ -4932,14 +4932,33 @@ void net_if_call_link_cb(struct net_if *iface, struct net_linkaddr *lladdr,
...
+	}
+
+	if (caps == ETHERNET_HW_RX_CHKSUM_OFFLOAD) {
+		config_type = ETHERNET_CONFIG_TYPE_RX_CHECKSUM_SUPPORT;
+	} else {
+		config_type = ETHERNET_CONFIG_TYPE_TX_CHECKSUM_SUPPORT;
+	}
+
+	if (net_eth_get_hw_config(iface, config_type, &config) != 0) {
+		return false; /* No extra info, assume all offloaded. */
```
Comment (pdgendt) [inline subsys/net/ip/net_if.c:4957]: Doesn't this apply to most in-tree drivers that had `true` before?

File: include/zephyr/net/ethernet.h (hunk 22 above)
Comment (WorldofJARcraft) [reply include/zephyr/net/ethernet.h]: Done.

**Review by pdgendt — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Review by decsny — DISMISSED** [review]
blocking until question is answered

**Comment by WorldofJARcraft** [comment]
> It feels kind of strange for this to come from the ethernet config, since this is not really a configuration but rather a hardware capability? why not come from the get_capabilities function?

The original motivation was to save bits in the existing hardware capabilities enum.
However, I think including checksum offloading in the configuration API actually makes more sense: For some devices, checksum offloading is a run-time configuration. For example, for the Xilinx AXI Ethernet, TX checksum offloading must be requested for each packet via the control stream / the "userapp" fields in the AXI DMA SG descriptor. If TX offloading is not requested, the NIC will gladly send a packet with no checksum. Integrating checksum offloading with the configuration interfaces makes it possible to enable and disable the feature and query whether it is currently active.

**Review by jukkar — APPROVED** [review]
